### PR TITLE
feat(HybridSolidInput): Added the grdecl input for hybrid solids.

### PR DESCRIPTION
**Pull request by BenPinet**

### CODE REVIEW

File: src/geode/geosciences_io/mesh/grdecl_input.cpp [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+                        geode::Point3D( { geode::string_to_double( tokens[0] ),
+                            geode::string_to_double( tokens[1] ),
+                            geode::string_to_double( tokens[2] ) } );
+                    geode::Point3D bottom_point =
+                        geode::Point3D( { geode::string_to_double( tokens[3] ),
+                            geode::string_to_double( tokens[4] ),
+                            geode::string_to_double( tokens[5] ) } );
+                    pillars.emplace_back(
+                        std::move( top_point ), std::move( bottom_point ) );
+                    std::getline( file_, line );
```
Comment (MelchiorSchuh) [inline src/geode/geosciences_io/mesh/grdecl_input.cpp:140]: put the getline directly after the if and remove the else condition

File: src/geode/geosciences_io/mesh/grdecl_input.cpp [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+            {
+                // DEBUG( line );
+                const auto tokens = geode::string_split( line );
+
+                if( tokens.size() != 0 )
+                {
+                    geode::Point3D top_point =
+                        geode::Point3D( { geode::string_to_double( tokens[0] ),
+                            geode::string_to_double( tokens[1] ),
+                            geode::string_to_double( tokens[2] ) } );
```
Comment (MelchiorSchuh) [inline src/geode/geosciences_io/mesh/grdecl_input.cpp:133]: create point directly ( geode::Point3D top_point{{x,y,z}}; )

File: src/geode/geosciences_io/mesh/grdecl_input.cpp [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+            geode::index_t depths_number{ 0 };
+            std::getline( file_, line );
+            while( line != "/" )
+            {
+                const auto tokens = geode::string_split( line );
+                for( const auto depths_id : geode::Indices{ tokens } )
+                {
+                    depths[depths_number++] =
+                        geode::string_to_double( tokens[depths_id] );
+                    depths_number++;
```
Comment (MelchiorSchuh) [inline src/geode/geosciences_io/mesh/grdecl_input.cpp:163]: remove the second one

File: src/geode/geosciences_io/mesh/grdecl_input.cpp [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+                }
+
+                std::getline( file_, line );
+            }
+            return depths;
+        }
+
+        std::array< geode::Point3D, 8 > get_cell_points(
+            const std::array< geode::index_t, 3 >& grid_coordinates,
+            const absl::FixedArray< double >& depths,
```
Comment (MelchiorSchuh) [inline src/geode/geosciences_io/mesh/grdecl_input.cpp:173]: Use Span

File: src/geode/geosciences_io/mesh/grdecl_input.cpp [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        Pillar( geode::Point3D top_point, geode::Point3D bottom_point )
+            : top{ std::move( top_point ) }, bottom{ std::move( bottom_point ) }
+        {
+        }
+
+        const geode::Point3D top;
+        const geode::Point3D bottom;
+    };
+
+    geode::Point3D interpolate_on_pillar(
```
Comment (MelchiorSchuh) [inline src/geode/geosciences_io/mesh/grdecl_input.cpp:52]: better naming

File: src/geode/geosciences_io/mesh/grdecl_input.cpp [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+                pillars[grid_coordinates[0]
+                        + ( nx_ + 1 ) * ( grid_coordinates[1] )];
+            const auto& bottom_right_pillar =
+                pillars[grid_coordinates[0]
+                        + ( nx_ + 1 ) * ( grid_coordinates[1] + 1 ) + 1];
+            const auto& top_right_pillar =
+                pillars[grid_coordinates[0]
+                        + ( nx_ + 1 ) * ( grid_coordinates[1] ) + 1];
+            std::array< geode::Point3D, 8 > points;
+            points[0] = interpolate_on_pillar(
```
Comment (MelchiorSchuh) [inline src/geode/geosciences_io/mesh/grdecl_input.cpp:189]: builder.create_point(interpolate...)

File: tests/mesh/test-grdecl.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+#include <geode/geosciences_io/mesh/private/grdecl_input.h>
+#include <geode/mesh/core/geode/geode_hybrid_solid.h>
+#include <geode/mesh/core/hybrid_solid.h>
+#include <geode/mesh/io/hybrid_solid_input.h>
+#include <geode/mesh/io/hybrid_solid_output.h>
+
+#include <geode/io/mesh/common.h>
+
+void check_solid( geode::HybridSolid3D& solid, geode::index_t nb_polyhedra )
```
Comment (MelchiorSchuh) [inline tests/mesh/test-grdecl.cpp:37]:
```suggestion
void check_solid( const geode::HybridSolid3D& solid, geode::index_t nb_polyhedra )
```

File: tests/mesh/test-grdecl.cpp [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+{
+    OPENGEODE_EXCEPTION( solid.nb_polyhedra() == nb_polyhedra,
+        "Number of polyhedra in the GrdeclHybridSolid is not correct" );
+}
+
+void check_file( absl::string_view filename, geode::index_t nb_polyhedra )
+{
+    // Load File
+    auto solid = geode::load_hybrid_solid< 3 >( filename );
```
Comment (MelchiorSchuh) [inline tests/mesh/test-grdecl.cpp:46]:
```suggestion
    const auto solid = geode::load_hybrid_solid< 3 >( filename );
```

File: src/geode/geosciences_io/mesh/grdecl_input.cpp [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+            }
+            return depths;
+        }
+
+        std::array< geode::Point3D, 8 > cell_points(
+            const std::array< geode::index_t, 3 >& grid_coordinates,
+            absl::Span< const Pillar > pillars,
+            absl::Span< const double > depths ) const
+        {
+            const auto& pillars_id = cell_pillars_id( grid_coordinates );
```
Comment (BotellaA) [inline src/geode/geosciences_io/mesh/grdecl_input.cpp:150]:
```suggestion
            const auto pillars_id = cell_pillars_id( grid_coordinates );
```

File: src/geode/geosciences_io/mesh/grdecl_input.cpp [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+            std::array< geode::Point3D, 8 > points;
+            points[0] = interpolate_on_pillar(
+                depths[2 * grid_coordinates[0] + 4 * nx_ * grid_coordinates[1]
+                       + 4 * nx_ * ny_ * ( grid_coordinates[2] + 1 ) + 2 * nx_],
+                bottom_left_pillar );
+            points[1] = ( interpolate_on_pillar(
+                depths[2 * grid_coordinates[0] + 4 * nx_ * grid_coordinates[1]
+                       + 4 * nx_ * ny_ * ( grid_coordinates[2] + 1 ) + 2 * nx_
+                       + 1],
+                bottom_right_pillar ) );
```
Comment (BotellaA) [inline src/geode/geosciences_io/mesh/grdecl_input.cpp:164]:
```suggestion
            points[1] = interpolate_on_pillar(
                depths[2 * grid_coordinates[0] + 4 * nx_ * grid_coordinates[1]
                       + 4 * nx_ * ny_ * ( grid_coordinates[2] + 1 ) + 2 * nx_
                       + 1],
                bottom_right_pillar );
```
to do for all points

File: src/geode/geosciences_io/mesh/grdecl_input.cpp [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+                       + 1 + 4 * nx_ * ny_ * grid_coordinates[2]],
+                top_right_pillar );
+            points[7] = interpolate_on_pillar(
+                depths[2 * grid_coordinates[0] + 4 * nx_ * grid_coordinates[1]
+                       + 4 * nx_ * ny_ * grid_coordinates[2]],
+                top_left_pillar );
+            return points;
+        }
+
+        geode::NNSearch3D::ColocatedInfo create_all_points(
```
Comment (BotellaA) [inline src/geode/geosciences_io/mesh/grdecl_input.cpp:192]:
```suggestion
        geode::NNSearch3D::ColocatedInfo create_points(
```

File: src/geode/geosciences_io/mesh/grdecl_input.cpp [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+                geode::NNSearch3D{ points }.colocated_index_mapping(
+                    geode::global_epsilon );
+            for( const auto& point : collocated_mapping.unique_points )
+            {
+                builder_->create_point( point );
+            }
+            return collocated_mapping;
+        }
+
+        void create_grid( absl::Span< const Pillar > pillars,
```
Comment (BotellaA) [inline src/geode/geosciences_io/mesh/grdecl_input.cpp:225]:
```suggestion
        void create_cells( absl::Span< const Pillar > pillars,
```

File: src/geode/geosciences_io/mesh/grdecl_input.cpp [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+            builder_->compute_polyhedron_adjacencies();
+        }
+
+        std::array< geode::index_t, 4 > cell_pillars_id(
+            const std::array< geode::index_t, 3 >& grid_coordinates ) const
+        {
+            const geode::index_t number_of_vertices_on_lines = nx_ + 1;
+            const geode::index_t bottom_left_pillar_id =
+                grid_coordinates[0]
+                + ( number_of_vertices_on_lines ) * ( grid_coordinates[1] + 1 );
```
Comment (BotellaA) [inline src/geode/geosciences_io/mesh/grdecl_input.cpp:252]:
```suggestion
                + number_of_vertices_on_lines * ( grid_coordinates[1] + 1 );
```
simplify readability

File: tests/CMakeLists.txt [hunk 14]
```diff
@@ -23,6 +23,7 @@ cmake_minimum_required(VERSION 3.15)
 if(NOT TARGET OpenGeode-GeosciencesIO::mesh)
     project(OpenGeode-GeosciencesIO CXX)
     find_package(OpenGeode REQUIRED CONFIG)
+    find_package(OpenGeode-IO REQUIRED CONFIG)
```
Comment (BotellaA) [inline tests/CMakeLists.txt:26]:
```suggestion
```
No

File: tests/mesh/CMakeLists.txt [hunk 15]
```diff
@@ -19,10 +19,11 @@
 # SOFTWARE.
 
 add_geode_test(
-    SOURCE "test-ts.cpp"
+    SOURCE "test-grdecl.cpp"
     DEPENDENCIES
         OpenGeode::basic
         OpenGeode::mesh
+        OpenGeode-IO::mesh
```
Comment (BotellaA) [inline tests/mesh/CMakeLists.txt:26]:
```suggestion
```

File: tests/mesh/test-grdecl.cpp [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+#include <geode/basic/assert.h>
+#include <geode/basic/logger.h>
+
+#include <geode/geosciences_io/mesh/private/grdecl_input.h>
+#include <geode/mesh/core/geode/geode_hybrid_solid.h>
+#include <geode/mesh/core/hybrid_solid.h>
+#include <geode/mesh/io/hybrid_solid_input.h>
+#include <geode/mesh/io/hybrid_solid_output.h>
+
+#include <geode/io/mesh/common.h>
```
Comment (BotellaA) [inline tests/mesh/test-grdecl.cpp:35]:
```suggestion
```

File: tests/mesh/test-grdecl.cpp [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    const auto solid = geode::load_hybrid_solid< 3 >( filename );
+    check_solid( *solid, nb_polyhedra );
+    geode::save_hybrid_solid( *solid, "toto.vtu" );
+}
+
+int main()
+{
+    try
+    {
+        geode::IOMeshLibrary::initialize();
```
Comment (BotellaA) [inline tests/mesh/test-grdecl.cpp:56]:
```suggestion
```

**Review by BotellaA — CHANGES_REQUESTED** [review]

**Review by BotellaA — APPROVED** [review]

**Pull request merged**